### DISCUSSION TITLE
RTMP camera live streaming setup

### DISCUSSION THREAD

**Discussion by chp77** [pedroSG94/RootEncoder#1287] [Q&A]
Hi all, I'm a newbie and want to ask the way how to do the live streaming via rtmp. Below is my code in the service:
.....
```
import com.pedro.library.rtmp.RtmpCamera1;
import com.pedro.rtmp.utils.ConnectCheckerRtmp;
import com.pedro.encoder.input.video.CameraOpenException;

public class USBCameraService extends Service {
    private static final String TAG = "USBCameraService";
    private final String RTMP_SERVER_URL = "rtmp://192.x.x.xx/live/run1";

    private MediaCodec mediaCodec;
    private Surface encoderInputSurface;

    private final Handler handler = new Handler(Looper.getMainLooper());
    private Runnable cameraCheckRunnable;
    private TextureView textureView;
    private RelativeLayout relativeLayout;
    private ImageView overlayImageView;

    private CameraManager cameraManager;
    private CameraDevice cameraDevice;
    private CameraCaptureSession cameraCaptureSession;
    private CaptureRequest.Builder previewRequestBuilder;
    private Size previewSize;
    private int desiredWidth = 640; // Set your desired width here
    private int desiredHeight = 360; // Set your desired height here
    private Button endButton; // The "End" button
    private boolean streamingEnabled = true;
    private OkHttpClient httpClient;
    private ImageReader imageReader;
    private byte[] frameData;
    private String uuid = "";
    private RtmpCamera1 rtmpCamera1;
    private ConnectCheckerRtmp connectCheckerRtmp;

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        super.onCreate();

        if (intent != null) {
            // Retrieve the text from the Intent extra
            uuid = intent.getStringExtra("uuid");
        }

        // Initialize the camera manager
        cameraManager = (CameraManager) getSystemService(Context.CAMERA_SERVICE);

        // Create a RelativeLayout to hold the views
        RelativeLayout relativeLayout = new RelativeLayout(this);

        // Initialize the "End" button
        endButton = new Button(this);
        endButton.setText("End");
        endButton.setTextColor(Color.WHITE);
        endButton.setBackgroundColor(Color.RED);

        // Set layout parameters for the "End" button
        RelativeLayout.LayoutParams buttonParams = new RelativeLayout.LayoutParams(
            200,
            80
        );

        buttonParams.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM); // Align to the bottom of the parent
        buttonParams.addRule(RelativeLayout.ALIGN_PARENT_END); // Align to the right end of the parent
        endButton.setLayoutParams(buttonParams);

        // Set border radius (rounded corners)
        GradientDrawable drawable = new GradientDrawable();
        drawable.setShape(GradientDrawable.RECTANGLE);
        drawable.setCornerRadius(10); // Set the corner radius in pixels
        drawable.setColor(Color.RED); // Set the button's background color
        endButton.setBackground(drawable);
        endButton.setTextSize(10);
        endButton.setPadding(20, 20, 20, 20);
        // Make the button clickable
        endButton.setClickable(true);
        endButton.setFocusable(true);

        // Inflate the layout
        LayoutInflater inflater = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View rootView = inflater.inflate(R.layout.activity_camera_preview, null);

        // Find the TextureView within the inflated layout
        textureView = (TextureView) rootView.findViewById(R.id.camera_preview);

        // Check if the TextureView already has a parent and remove it if necessary
        if (textureView.getParent() != null) {
            ((ViewGroup) textureView.getParent()).removeView(textureView);
        }

        // Set layout parameters for the TextureView
        RelativeLayout.LayoutParams textureParams = new RelativeLayout.LayoutParams(
            desiredWidth, // Set the width here
            desiredHeight // Set the height here
        );
        textureParams.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM); // Align to the bottom of the parent
        textureParams.addRule(RelativeLayout.ALIGN_PARENT_END); // Align to the right end of the parent
        textureView.setLayoutParams(textureParams);

        // Add the "End" button and the TextureView to the RelativeLayout
        relativeLayout.addView(textureView);
        relativeLayout.addView(endButton);

        endButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Log.e("end btn clicked", "true");
                try {
                    WindowManager windowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
                    windowManager.removeView(relativeLayout);
                    closeCamera();
                } catch (Exception e) {
                    // Log any exceptions that occur
                    Log.e("View Removal Error", "Error removing the view: " + e.getMessage());
                }
            }
        });

        // Get the screen dimensions
        DisplayMetrics displayMetrics = new DisplayMetrics();
        WindowManager windowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
        windowManager.getDefaultDisplay().getMetrics(displayMetrics);

        int screenWidth = displayMetrics.widthPixels;
        int screenHeight = displayMetrics.heightPixels;

        // Calculate the position for the TextureView (bottom-right corner)
        int xPosition = screenWidth - desiredWidth; // Right edge of the screen
        int yPosition = screenHeight - desiredHeight; // Bottom edge of the screen

        // Create layout parameters for the view
        WindowManager.LayoutParams params = new WindowManager.LayoutParams(
            desiredWidth, // Set the width here
            desiredHeight, // Set the height here
            xPosition, // Set the x position
            yPosition, // Set the y position
            WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
            PixelFormat.TRANSLUCENT
        );

        // Add the RelativeLayout to the WindowManager
        windowManager.addView(relativeLayout, params);

        rtmpCamera1 = new RtmpCamera1(textureView, connectCheckerRtmp);
        rtmpCamera1.setReTries(10);
        rtmpCamera1.startStream("rtmp://192.x.x.xxx/live/run1");

        checkForUSBCamera();

        return START_STICKY;
    }
```
....
can anyone assist me, really appreciate it

**Comment by pedroSG94** [comment]
Hello,

You have an example here to stream using a service:
https://github.com/pedroSG94/RootEncoder/blob/master/app/src/main/java/com/pedro/streamer/backgroundexample/RtpService.kt

According with your code you are using usb camera to stream so you will need change the code a bit. In your case you will need use RtmpClient directly and send video frames in h264 to it.

Remember that you will need send video info to RtmpClient (sps and pps) to work.

**Reply by chp77** [reply]
This is the version 2.2.8??
I encounter errors when trying to follow the reference above...


BUILD FAILED in 3s

Error: Command failed: ./gradlew app:installDebug -PreactNativeDevServerPort=8081
/home/chp/AwesomeProject/android/app/src/main/java/com/awesomeproject/USBCameraService.java:60: error: package com.pedro.streamer does not exist
import com.pedro.streamer.R;
                         ^
/home/chp/AwesomeProject/android/app/src/main/java/com/awesomeproject/USBCameraService.java:356: error: cannot find symbol
    private final ConnectCheckerRtp connectCheckerRtp = new ConnectCheckerRtp() {
                  ^
  symbol:   class ConnectCheckerRtp
  location: class USBCameraService
/home/chp/AwesomeProject/android/app/src/main/java/com/awesomeproject/USBCameraService.java:356: error: cannot find symbol
    private final ConnectCheckerRtp connectCheckerRtp = new ConnectCheckerRtp() {
                                                            ^
  symbol:   class ConnectCheckerRtp
  location: class USBCameraService
/home/chp/AwesomeProject/android/app/src/main/java/com/awesomeproject/USBCameraService.java:358: error: method does not override or implement a method from a supertype
        @Override
        ^
/home/chp/AwesomeProject/android/app/src/main/java/com/awesomeproject/USBCameraService.java:363: error: method does not override or implement a method from a supertype
        @Override
        ^
/home/chp/AwesomeProject/android/app/src/main/java/com/awesomeproject/USBCameraService.java:369: error: method does not override or implement a method from a supertype
        @Override
        ^
/home/chp/AwesomeProject/android/app/src/main/java/com/awesomeproject/USBCameraService.java:374: error: method does not override or implement a method from a supertype
        @Override
        ^
/home/chp/AwesomeProject/android/app/src/main/java/com/awesomeproject/USBCameraService.java:380: error: method does not override or implement a method from a supertype
        @Override
        ^
/home/chp/AwesomeProject/android/app/src/main/java/com/awesomeproject/USBCameraService.java:385: error: method does not override or implement a method from a supertype
        @Override
        ^
/home/chp/AwesomeProject/android/app/src/main/java/com/awesomeproject/USBCameraService.java:390: error: method does not override or implement a method from a supertype
        @Override
        ^
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
10 errors

**Comment by pedroSG94** [comment]
Hello,

You have multiple obvious errors.
```kotlin
import com.pedro.streamer.R;

```
This is the resource import from my project. This should be removed and replace code that use it.

ConnectCheckerRtp class is a class created for the example to join ConnectCheckerRtmp and ConnectCheckerRtsp in one.
You can replace it with one of that classes or get the class here:
https://github.com/pedroSG94/RootEncoder/blob/master/app/src/main/java/com/pedro/streamer/backgroundexample/ConnectCheckerRtp.kt

About your requirements using UVC camera. Maybe you can try this project (the code is old but you can try update to last library version):
https://github.com/pedroSG94/Stream-USB-test
This is unmaintained but you can check it
